code/modules/mob/living/living_fov.dm

```diff
@@ -0,0 +1,262 @@
+	var/x_shift = view_size[1] - current_fov_x
+	var/y_shift = view_size[2] - current_fov_y
+	var/x_scale = view_size[1] / current_fov_x
+	var/y_scale = view_size[2] / current_fov_y
+	current_fov_x = view_size[1]
+	current_fov_y = view_size[2]
+	visual_shadow.transform = blocker_mask.transform = new_matrix.Scale(x_scale, y_scale)
+	visual_shadow.transform = blocker_mask.transform = new_matrix.Translate(x_shift * 16, y_shift * 16)
+
+/// Adds the masks to the user
+/datum/field_of_view/proc/add_mask()
+	var/client/parent_client = owner.client
+	if(!parent_client) //Love client volatility!!
+		return
+	parent_client.screen += blocker_mask
+	parent_client.screen += visual_shadow
+	applied_mask = TRUE
+
+/// Removes the masks from the user
+/datum/field_of_view/proc/remove_mask()
+	var/client/parent_client = owner.client
+	if(!parent_client) //Love client volatility!!
+		return
+	parent_client.screen -= blocker_mask
+	parent_client.screen -= visual_shadow
+	applied_mask = FALSE
+
+/// When a direction of the user changes, so do the masks
+/datum/field_of_view/proc/dir_change(new_dir)
+	blocker_mask.dir = new_dir
+	visual_shadow.dir = new_dir
+
+/datum/field_of_view/proc/set_fov_angle(new_angle)
+	fov_angle = new_angle
+	blocker_mask.icon_state = "[fov_angle]"
+	visual_shadow.icon_state = "[fov_angle]_v"
+
+/// Is `observed_atom` in a mob's field of view? This takes blindness, nearsightness and FOV into consideration
+/mob/living/proc/in_fov(atom/observed_atom, ignore_self = FALSE)
+	if(ignore_self && observed_atom == src)
+		return TRUE
+	if(is_blind())
+		return FALSE
+	. = FALSE
+	var/turf/my_turf = get_turf(src) //Because being inside contents of something will cause our x,y to not be updated
+	// If turf doesn't exist, then we wouldn't get a fov check called by `play_fov_effect` or presumably other new stuff that might check this.
+	//  ^ If that case has changed and you need that check, add it.
+	var/rel_x = observed_atom.x - my_turf.x
+	var/rel_y = observed_atom.y - my_turf.y
+	if(fov_view)
+		if(rel_x >= -1 && rel_x <= 1 && rel_y >= -1 && rel_y <= 1) //Cheap way to check inside that 3x3 box around you
+			return TRUE //Also checks if both are 0 to stop division by zero
+	
+		// Get the vector length so we can create a good directional vector
+		var/vector_len = sqrt(abs(rel_x) ** 2 + abs(rel_y) ** 2)
+	
+		/// Getting a direction vector
+		var/dir_x
+		var/dir_y
+		switch(dir)
+			if(SOUTH)
+				dir_x = 0
+				dir_y = -vector_len
+			if(NORTH)
+				dir_x = 0
+				dir_y = vector_len
+			if(EAST)
+				dir_x = vector_len
+				dir_y = 0
+			if(WEST)
+				dir_x = -vector_len
+				dir_y = 0
+	
+		///Calculate angle
+		var/angle = arccos((dir_x * rel_x + dir_y * rel_y) / (sqrt(dir_x**2 + dir_y**2) * sqrt(rel_x**2 + rel_y**2)))
+	
+		/// Calculate vision angle and compare
+		var/vision_angle = (360 - fov_view) / 2
+		if(angle < vision_angle)
+			. = TRUE
+	else
+		. = TRUE
+
+	// Handling nearsightnedness
+	if(. && HAS_TRAIT(src, TRAIT_NEARSIGHT))
```

Comment:
this should probably be an early return

Reply:
I dont think there should be? There are checks we have to run to determine whether someone is really nearsighted, and if that person is far away to receive those effects

Reply:
```suggestion
	if(!. || !HAS_TRAIT(src, TRAIT_NEARSIGHT))
	    return
	... the rest of the code but on one less indentation level
```

Reply:
I suppose it can but wouldn't that be less readable?

Reply:
It shouldn't be an early-return, early-returns are for conditionals that are required for the process to work, generally. Having a lot of different conditionals not early-returned is fine, it means it's easy to add more.